airlock/management/commands/run_file_uploader.py

```diff
@@ -0,0 +1,146 @@
+from airlock.business_logic import bll
+from airlock.enums import RequestStatus
+from airlock.users import User
+
+
+logger = logging.getLogger(__name__)
+
+# We need a user with the output-checker role to access some
+# bll methods
+system_user = User("system", output_checker=True)
+
+
+class Command(BaseCommand):
+    """
+    Poll database for approved requests that still have files pending upload
+    """
+
+    def add_arguments(self, parser):
+        # In production, we want this check to run forever. Using a
+        # function means that we can test it on a finite number of loops.
+        parser.add_argument("--run-fn", default=lambda: True)
+
+    def handle(self, *args, **options):
+        run_fn = options["run_fn"]
+
+        logger.debug("Watching for tasks")
+
+        while run_fn():  # pragma: no branch
+            # Find approved requests
+            approved_requests = bll.get_approved_requests(user=system_user)
+
+            if not approved_requests:
+                # No pending file uploads found; wait for UPLOAD_DELAY seconds
+                # before checking again
+                time.sleep(settings.UPLOAD_DELAY)
+                continue
+
+            for approved_request in approved_requests:
+                # Find incomplete file uploads for this request
+                # This retrieves ALL incomplete file uploads, including those that have
+                # been reached the max attempt limit, so we can check for any
+                # requests that should be set to released
+                files_for_upload = get_upload_files_and_update_request_status(
+                    approved_request
+                )
+                if not files_for_upload:
+                    # All files are now uploaded, the status has been updated to released,
+                    # nothing to do
+                    continue
+
+                workspace = bll.get_workspace(approved_request.workspace, system_user)
+                for file_for_upload in files_for_upload:
+                    if file_for_upload.upload_attempts >= settings.UPLOAD_MAX_ATTEMPTS:
+                        logger.warning(
+                            "Max upload attempts reached for %s - %s, skipping",
+                            approved_request.id,
+                            file_for_upload.relpath,
+                        )
+                        continue
+                    # increment the retry attempts; if something goes wrong, we still
+                    # want this to be updated
+                    file_for_upload = bll.register_file_upload_attempt(
+                        approved_request, file_for_upload.relpath
+                    )
+                    try:
+                        do_upload_task(file_for_upload, approved_request, workspace)
+                    except Exception as error:
+                        # Catch any unexpected exception here so we don't stop the task runner
+                        # from running
```

Comment:
This seems right, but it would be good to be able to get notified of errors if they do happen. Is there some way we can (waves hands) use Honeycomb to do this?

Reply:
Oh yes, actually adding telemetry to this in general was another to-do that I didn't note in the PR description